### PR TITLE
Preliminary & experimental support for... [ci: last-only]

### PR DESCRIPTION
**Pull request by adriaanm**
... JPMS module access checks :-)

This will only be available in nightly builds (produced on jdk11), under -Yjpms. This also shows a path for future jdk11-specific features and how we might ship them before we bump our required Java version.

The PR is in rough shape, but it works for simple stuff. Putting this out there to facilitate discussion.

@retronym did most of the work, I just refactored until my paranoia of not affecting normal compiler operation was satisfied.

We wanted to have the basics in place before RC1, but we won't have time to refine or develop new functionality until later in the 2.13.x series.

### CODE REVIEW

**Comment by adriaanm** [comment]
Test Result (8 failures / +3)
scala.lang.stringinterpol.StringContextTest.initializationError
scala.tools.nsc.symtab.FreshNameExtractorTest.initializationError
scala.tools.testing.AssertThrowsTest.initializationError
test.files.run_.global-showdef.scala
test.files.run_.t6240-universe-code-gen.scala
test.files.run_.t6502.scala
test.files.run_.t8015-ffc.scala
test.macro-annot.run_.repl

**Comment by smarter** [comment]
How come all of scala.reflect.internal.jpms is written in Java and not Scala ?

**Comment by retronym** [comment]
I wanted to make it possible to host that code in an external build, built
with Java (9) only like Scala-asm, and then incorporate it into
Scala-compiler.jar

File: src/library/scala/collection/StringOps.scala [hunk 1]
```diff
@@ -628,9 +628,12 @@ final class StringOps(private val s: String) extends AnyVal {
     *  end characters; i.e., apply `.stripLineEnd` to all lines
     *  returned by `linesWithSeparators`.
     */
-  def lines: Iterator[String] =
+  def linesIterator: Iterator[String] =
     linesWithSeparators map (_.stripLineEnd)
 
+  @deprecated("Use linesIterator. JDK 11 introduces its own lines method on String", "2.13.0-M5")
```
Comment (xuwei-k) [inline src/library/scala/collection/StringOps.scala:634]: `2.13.0-RC1` or `2.13.0` ?

Comment (adriaanm) [reply src/library/scala/collection/StringOps.scala:634]: Yes, should be RC1 now :-) I'll move these commits to their own PR.

Comment (SethTisue) [reply src/library/scala/collection/StringOps.scala:634]: we normally just put "2.13.0" so we don't have to keep changing it

File: src/compiler-jdk11/scala/tools/nsc/jpms/JpmsClassPath.scala [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    }
+  }
+
+  override def asURLs: Seq[URL] = Nil // TODO
+  override def asClassPathStrings: Seq[String] = Nil // TODO
+  override def asSourcePathString: String = ""
+
+  private val fileManager = impl.getFileManager
+  private def locationAsPaths(location: StandardLocation): IndexedSeq[Path] = getLocationAsPaths(fileManager, location).asScala.toVector
+  private val packageIndex = mutable.AnyRefMap[String, PackageEntryImpl]()
```
Comment (retronym) [inline src/compiler-jdk11/scala/tools/nsc/jpms/JpmsClassPath.scala:71]: Unused.

File: src/compiler-jdk11/scala/tools/nsc/jpms/JpmsClassPath.scala [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+          if (Files.isDirectory(file)) {
+            val fileName = file.getFileName.toString
+            if (!seen(fileName) && !fileName.contains("-")) {
+              seen += fileName
+              val name = if (inPackage == "") fileName else inPackage + "." + fileName
+              result += PackageEntryImpl(name)
+            }
+          }
+        }
+      }
```
Comment (retronym) [inline src/compiler-jdk11/scala/tools/nsc/jpms/JpmsClassPath.scala:122]: We need an `else` branch for JARs on the module path, for which `path` is the `.jar` file.

I think the right approach is to index the package structure of the JARs eagerly, then use that index here.


WIP: https://github.com/retronym/scala/tree/review/jdk11

The indexing in that branch could be improved to be hierarchical.

To test:

```
$ cat sandbox/test.scala
class Test { println(org.apache.commons.io.IOUtils.EOF) }

$ ./build/quick/bin/scala -nobootcp -Yjpms -modulepath $(coursier fetch -q -p commons-io:commons-io:2.6) -addmodules:org.apache.commons.io -addreads:org.apache.commons.io=ALL-UNNAMED
```

**Comment by retronym** [comment]
We should make it possible for tools like Zinc to see inside `PlainNioFile` to get at the underlying `nio.file.Path`. We should also implement `PlainNioFile.underlyingSource` to get to the underlying JAR file.

Sketch:

```
scala> def pathOfNioPath(f: scala.reflect.io.AbstractFile) = { val fld = f.getClass.getDeclaredField("nioPath"); fld.setAccessible(true); fld.get(f).asInstanceOf[java.nio.file.Path] }
pathOfNioPath: (f: scala.reflect.io.AbstractFile)java.nio.file.Path

scala> val assocFile = rootMirror.getRequiredClass("org.apache.commons.io.IOUtils").initialize.associatedFile
assocFile: scala.reflect.io.AbstractFile = /org/apache/commons/io/IOUtils.class

scala> val assocNioPath = pathOfNioPath(assocFile)
assocNioPath: java.nio.file.Path = /org/apache/commons/io/IOUtils.class

scala> assocNioPath.getFileSystem.provider.getScheme
res74: String = jar

scala> assocNioPath.getFileSystem.getFileStores.iterator.next.name
res83: String = /Users/jz/.coursier/cache/v1/https/repo1.maven.org/maven2/commons-io/commons-io/2.6/commons-io-2.6.jar/
```

**Comment by som-snytt** [comment]
Kind of makes you miss `-Xexperimental`.

**Comment by som-snytt** [comment]
Thanks, this open PR was the only reason I remembered how to say `[ci: last-only]`.

Also, a medium-sized spider, kind of brownish-yellow and translucent, just came all the way down from the ceiling on a thread to land in the detritus on my computer table. I thought I would wind up squashing it and then feel bad, but now it seems to have scurried off to a corner where I will probably squash it unwittingly and feel OK.

For the record, at the Java office, I had to push to get onto Java 8, then I was pushing again to get on the release train because by the time it drops, you haven't learned what you need to learn. Now we're on Java 11, more or less. The fellow in the next cubicle said he found a bug in open jdk overload resolution, because it works in Eclipse but fails in the gradle build. I said I'd be very surprised. I wound up revisiting the Java spec on overloading for the first time since a Scala ticket in 2012. There must be a good sports metaphor for this phenomenon, where you're always trying to stretch it into a triple and you have to slide into the bag.

(That's baseball. I also meant to add that this spider was nothing like what they have in Australia, as I've read on the internet.)

**Comment by adriaanm** [comment]
Itsy-bitsy will rise again.

**Pull request closed without merging**